pymc3/distributions/multivariate.py

```diff
@@ -93,31 +77,52 @@ class MvNormal(Continuous):
 
     """
 
-    def __init__(self, mu, cov=None, tau=None, gpu_compat=False, *args, **kwargs):
+    def __init__(self, mu, cov=None, tau=None, chol=None, gpu_compat=False,
+                 *args, **kwargs):
         super(MvNormal, self).__init__(*args, **kwargs)
         self.mean = self.median = self.mode = self.mu = mu = tt.as_tensor_variable(mu)
-        tau, cov = get_tau_cov(mu, tau=tau, cov=cov)
+        tau, chol = get_tau_chol(tau=tau, cov=cov, chol=chol)
         self.tau = tt.as_tensor_variable(tau)
-        self.cov = tt.as_tensor_variable(cov)
+        self.chol = tt.as_tensor_variable(chol)
+        # not necessary to define theano solver function if tau given...
+        self.solve = tt.slinalg.Solve(A_structure="lower_triangular", lower=True)
         self.gpu_compat = gpu_compat
         if gpu_compat is False and theano.config.device == 'gpu':
             warnings.warn("The function used is not GPU compatible. Please check the gpu_compat flag")
 
     def random(self, point=None, size=None):
-        mu, cov = draw_values([self.mu, self.cov], point=point)
+        mu, chol = draw_values([self.mu, self.chol], point=point)
 
-        def _random(mean, cov, size=None):
-            return stats.multivariate_normal.rvs(
-                mean, cov, None if size == mean.shape else size)
+        if size is None or size == mu.shape:
+            size = mu.shape
+        elif isinstance(size, collections.Iterable):
+            size = list(size)
+            size.append(mu.shape[0])
+        else:
+            size = [size]
+            size.append(mu.shape[0])
 
-        samples = generate_samples(_random,
-                                   mean=mu, cov=cov,
-                                   dist_shape=self.shape,
-                                   broadcast_shape=mu.shape,
-                                   size=size)
-        return samples
+        return mu + (np.dot(np.random.standard_normal(size), chol))
 
     def logp(self, value):
+        if self.chol is not None:
+            return _logp_chol(self, value)
+        else:
+            return _logp_tau(self, value)
+
+    def _logp_chol(self, value):
+        mu = self.mu
+        chol = self.chol
+
+        delta = value - mu
+        k = chol.shape[0]
+        tmp = self.solve(chol, delta)
+
+        return k * tt.log(2 * np.pi) +\
```

Comment:
Thanks @ferrine !  I didn't know that was in there.  The goal is to not have direct inversions and determinant calculations.  These are not as numerically stable or fast as using Cholesky with solves.  This is why we have `_logp_tau` and `_logp_chol`.  Unfortunately, it looks like `log_normal_mv` does not use this.